### PR TITLE
Make s3 object metadata available on downloads

### PR DESCRIPTION
**Pull request by kag0**
independent from stream materialization ( #740)

* download method in s3 client now returns a tuple with stream and metadata
* stream from download now materializes into NotUsed

### CODE REVIEW

File: s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala [hunk 1]
```diff
@@ -68,19 +68,17 @@ private[alpakka] final class S3Stream(settings: S3Settings)(implicit system: Act
...
-          ObjectMetadata(resp.headers)
-        }
-      }
+    (
+      Source
+        .fromFuture(future.flatMap(entityForSuccess))
+        .map(_.dataBytes)
+        .flatMapConcat(identity),
+      future.map(resp ⇒ ObjectMetadata(resp.headers))
+    )
```
Comment (ennru) [inline s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala]: Took me a moment to see the tuple, wouldn't it be good to use intermediate vals?

File: s3/src/main/scala/akka/stream/alpakka/s3/javadsl/S3Client.scala [hunk 2]
```diff
@@ -421,49 +423,49 @@ final class S3Client(s3Settings: S3Settings, system: ActorSystem, mat: Materiali
...
+    JPair(stream.asJava, meta.map(metaDataToJava)(mat.executionContext).toJava)
+  }
+
   /**
    * Downloads a S3 Object
    *
    * @param bucket the s3 bucket name
    * @param key the s3 object key
-   * @return A [[Source]] of [[ByteString]] that materializes into a [[CompletionStage]] containing the [[ObjectMetadata]]
+   * @return A [[Pair]] with a [[Source]] of [[ByteString]], and a [[CompletionStage]] containing the [[ObjectMetadata]]
```
Comment (ennru) [inline s3/src/main/scala/akka/stream/alpakka/s3/javadsl/S3Client.scala]: Links to other classes almost always need to be fully qualified in Scaladoc. (But it's wront in many places in Alpakka.)

Comment (kag0) [reply s3/src/main/scala/akka/stream/alpakka/s3/javadsl/S3Client.scala]: added fully qualified `Pair`, let me know if you want to update `Source`, `CompletionStage`, etc as well.

Comment (ennru) [reply s3/src/main/scala/akka/stream/alpakka/s3/javadsl/S3Client.scala]: Yes, please.

**Comment by kag0** [comment]
@johanandren here's one with akka http as it might look if one were streaming a file from s3 back to a client
```scala
val (data, eventualMeta) = s3Client.download(???, ???)
complete( eventualMeta.map ( meta ⇒
    HttpResponse(
      entity = HttpEntity(
        meta.contentType.flatMap(ContentType.parse(_).toOption).getOrElse(`application/octet-stream`),
        meta.contentLength,
        data
      )
    )
) )
```

**Comment by ennru** [comment]
It would be good to see this kind of example in the documentation.

**Comment by kag0** [comment]
I'm not familiar with paradox, so you'll have to let me know if I did that right.

**Comment by ennru** [comment]
Thank you for your contribution! The S3 connector seems to be a popular way to get started using Alpakka.

**Pull request merged**